File: 14_Bowling/javascript/bowling.js
// BOWLING
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(34) + "BOWL\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    c = [];
    a = [];
    for (i = 0; i <= 15; i++)
        c[i] = 0;
    print("WELCOME TO THE ALLEY\n");
    print("BRING YOUR FRIENDS\n");
    print("OKAY LET'S FIRST GET ACQUAINTED\n");
    print("\n");
    print("THE INSTRUCTIONS (Y/N)\n");
    str = await input();
    if (str.substr(0, 1) == "Y") {
        print("THE GAME OF BOWLING TAKES MIND AND SKILL.DURING THE GAME\n");
        print("THE COMPUTER WILL KEEP SCORE.YOU MAY COMPETE WITH\n");
        print("OTHER PLAYERS[UP TO FOUR].YOU WILL BE PLAYING TEN FRAMES\n");
        print("ON THE PIN DIAGRAM 'O' MEANS THE PIN IS DOWN...'+' MEANS THE\n");
        print("PIN IS STANDING.AFTER THE GAME THE COMPUTER WILL SHOW YOUR\n");
        print("SCORES .\n");
    }
    print("FIRST OF ALL...HOW MANY ARE PLAYING");
    r = parseInt(await input());
    while (1) {
        print("\n");
        print("VERY GOOD...\n");
        for (i = 1; i <= 100; i++) {
            a[i] = [];
            for (j = 1; j <= 6; j++)
                a[i][j] = 0;
        }
        f = 1;
        do {
            for (p = 1; p <= r; p++) {
                // m = 0; // Repeated in original
                b = 1;
                m = 0;
                q = 0;
                for (i = 1; i <= 15; i++)
                    c[i] = 0;
                while (1) {
                    // Ball generator using mod '15' system
                    print("TYPE ROLL TO GET THE BALL GOING.\n");
                    ns = await input();
                    k = 0;
                    d = 0;
                    for (i = 1; i <= 20; i++) {
                        x = Math.floor(Math.random() * 100);
                        for (j = 1; j <= 10; j++)
                            if (x < 15 * j)
                                break;
                        c[15 * j - x] = 1;
                    }
                    // Pin diagram
                    print("PLAYER: " + p + " FRAME: " + f + " BALL: " + b + "\n");
                    print("\n");
                    for (i = 0; i <= 3; i++) {
                        str = "";
                        for (j = 1; j <= 4 - i; j++) {
                            k++;
                            while (str.length < i)
                                str += " ";
                            if (c[k] == 1)
                                str += "O ";
                            else
                                str += "+ ";
                        }
                        print(str + "\n");
                    }
                    // Roll analysis
                    for (i = 1; i <= 10; i++)
                        d += c[i];
                    if (d - m == 0)
                        print("GUTTER!!\n");
                    if (b == 1 && d == 10) {
                        print("STRIKE!!!!!\n");
                        q = 3;
                    }
                    if (b == 2 && d == 10) {
                        print("SPARE!!!!\n");
                        q = 2;
                    }
                    if (b == 2 && d < 10) {
                        print("ERROR!!!\n");
                        q = 1;
                    }
                    if (b == 1 && d < 10) {
                        print("ROLL YOUR 2ND BALL\n");
                    }
                    // Storage of the scores
                    print("\n");
                    a[f * p][b] = d;
                    if (b != 2) {
                        b = 2;
                        m = d;
                        if (q == 3) {
                            a[f * p][b] = d;
                        } else {
                            a[f * p][b] = d - m;
                            if (q == 0) // ROLL
                                continue;
                        }
                    }
                    break;
                }
                a[f * p][3] = q;
            }
        } while (++f < 11) ;
        print("FRAMES\n");
        for (i = 1; i <= 10; i++)
            print(" " + i + " ");
        print("\n");
        for (p = 1; p <= r; p++) {
            for (i = 1; i <= 3; i++) {
                for (j = 1; j <= 10; j++) {
                    print(" " + a[j * p][i] + " ");
                }
                print("\n");
            }
            print("\n");
        }
        print("DO YOU WANT ANOTHER GAME");
        str = await input();
        if (str.substr(0, 1) != "Y")
            break;
        // Bug in original game, jumps to 2610, without restarting P variable
    }
}

main();

File: 14_Bowling/java/Bowling.java
import java.util.Scanner;
import java.lang.Math;

/**
 * Game of Bowling
 * <p>
 * Based on the BASIC game of Bowling here
 * https://github.com/coding-horror/basic-computer-games/blob/main/14%20Bowling/bowling.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's BASIC game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 *
 * Converted from BASIC to Java by Darren Cardenas.
 */

public class Bowling {

  private final Scanner scan;  // For user input

  public Bowling() {

    scan = new Scanner(System.in);

  }  // End of constructor Bowling

  public void play() {

    showIntro();
    startGame();

  }  // End of method play

  private static void showIntro() {

    System.out.println(" ".repeat(33) + "BOWL");
    System.out.println(" ".repeat(14) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
    System.out.println("\n\n");

  }  // End of method showIntro

  private void startGame() {

    int ball = 0;
    int bell = 0;
    int frame = 0;
    int ii = 0;  // Loop iterator
    int jj = 0;  // Loop iterator
    int kk = 0;  // Loop iterator
    int numPlayers = 0;
    int pinsDownBefore = 0;
    int pinsDownNow = 0;
    int player = 0;
    int randVal = 0;
    int result = 0;

    int[] pins = new int[16];

    int[][] scores = new int[101][7];

    String userResponse = "";

    System.out.println("WELCOME TO THE ALLEY");
    System.out.println("BRING YOUR FRIENDS");
    System.out.println("OKAY LET'S FIRST GET ACQUAINTED");
    System.out.println("");
    System.out.println("THE INSTRUCTIONS (Y/N)");
    System.out.print("? ");

    userResponse = scan.nextLine();

    if (userResponse.toUpperCase().equals("Y")) {
      printRules();
    }

    System.out.print("FIRST OF ALL...HOW MANY ARE PLAYING? ");
    numPlayers = Integer.parseInt(scan.nextLine());

    System.out.println("");
    System.out.println("VERY GOOD...");

    // Begin outer while loop
    while (true) {

      for (ii = 1; ii <= 100; ii++) {
        for (jj = 1; jj <= 6; jj++) {
          scores[ii][jj] = 0;
        }
      }

      frame = 1;

      // Begin frame while loop
      while (frame < 11) {

        // Begin loop through all players
        for (player = 1; player <= numPlayers; player++) {

          pinsDownBefore = 0;
          ball = 1;
          result = 0;

          for (ii = 1; ii <= 15; ii++) {
            pins[ii] = 0;
          }

          while (true) {

            // Ball generator using mod '15' system

            System.out.println("TYPE ROLL TO GET THE BALL GOING.");
            System.out.print("? ");
            scan.nextLine();

            kk = 0;
            pinsDownNow = 0;

            for (ii = 1; ii <= 20; ii++) {

              randVal = (int)(Math.random() * 100) + 1;

              for (jj = 1; jj <= 10; jj++) {

                if (randVal < 15 * jj) {
                  break;
                }
              }
              pins[15 * jj - randVal] = 1;
            }

            // Pin diagram

            System.out.println("PLAYER: " + player + " FRAME: " + frame + " BALL: " + ball);

            for (ii = 0; ii <= 3; ii++) {

              System.out.println("");

              System.out.print(" ".repeat(ii));

              for (jj = 1; jj <= 4 - ii; jj++) {

                kk++;

                if (pins[kk] == 1) {

                  System.out.print("O ");

                } else {

                  System.out.print("+ ");
                }
              }
            }

            System.out.println("");

            // Roll analysis

            for (ii = 1; ii <= 10; ii++) {
              pinsDownNow += pins[ii];
            }

            if (pinsDownNow - pinsDownBefore == 0) {
              System.out.println("GUTTER!!");
            }

            if (ball == 1 && pinsDownNow == 10) {
              System.out.println("STRIKE!!!!!");

              // Ring bell
              for (bell = 1; bell <= 4; bell++) {
                System.out.print("\007");
                try {
                  Thread.sleep(500);
                } catch (InterruptedException e) {
                  Thread.currentThread().interrupt();
                }
              }
              result = 3;
            }

            if (ball == 2 && pinsDownNow == 10) {
              System.out.println("SPARE!!!!");
              result = 2;
            }

            if (ball == 2 && pinsDownNow < 10) {
              System.out.println("ERROR!!!");
              result = 1;
            }

            if (ball == 1 && pinsDownNow < 10) {
              System.out.println("ROLL YOUR 2ND BALL");
            }

            // Storage of the scores

            System.out.println("");

            scores[frame * player][ball] = pinsDownNow;

            if (ball != 2) {
              ball = 2;
              pinsDownBefore = pinsDownNow;

              if (result != 3) {
                scores[frame * player][ball] = pinsDownNow - pinsDownBefore;
                if (result == 0) {
                  continue;
                }
              } else {
                scores[frame * player][ball] = pinsDownNow;
              }

            }
            break;
          }

          scores[frame * player][3] = result;

        }  // End loop through all players

        frame++;

      }  // End frame while loop

      System.out.println("FRAMES");

      System.out.print(" ");
      for (ii = 1; ii <= 10; ii++) {
        System.out.print(ii + " ");
      }

      System.out.println("");

      for (player = 1; player <= numPlayers; player++) {
        for (ii = 1; ii <= 3; ii++) {
          System.out.print(" ");
          for (jj = 1; jj <= 10; jj++) {
            System.out.print (scores[jj * player][ii] + " ");
          }
          System.out.println("");
        }
        System.out.println("");
      }

      System.out.println("DO YOU WANT ANOTHER GAME");
      System.out.print("? ");

      userResponse = scan.nextLine();

      if (!String.valueOf(userResponse.toUpperCase().charAt(0)).equals("Y")) {
        break;
      }

    }  // End outer while loop

  }  // End of method startGame

  public static void printRules() {

    System.out.println("THE GAME OF BOWLING TAKES MIND AND SKILL.DURING THE GAME");
    System.out.println("THE COMPUTER WILL KEEP SCORE.YOU MAY COMPETE WITH");
    System.out.println("OTHER PLAYERS[UP TO FOUR].YOU WILL BE PLAYING TEN FRAMES");
    System.out.println("ON THE PIN DIAGRAM 'O' MEANS THE PIN IS DOWN...'+' MEANS THE");
    System.out.println("PIN IS STANDING.AFTER THE GAME THE COMPUTER WILL SHOW YOUR");
    System.out.println("SCORES .");

  }  // End of method printRules

  public static void main(String[] args) {

    Bowling game = new Bowling();
    game.play();

  }  // End of method main

}  // End of class Bowling
